### PR TITLE
[`flake8-use-pathlib`] Recommend `Path.iterdir()` over `os.listdir()` (`PTH208`)

### LINKED ISSUE #14490: New `flake8-pathlib` rule: `os.listdir` (PTH208)
The pathlib rules supported by `ruff` could be extended with detection with `os.listdir`:

```python
import os
from pathlib import Path

p = Path(".")
print(os.listdir(p))
```

use instead:

```python
from pathlib import Path

p = Path(".")
print(list(p.iterdir()))
```

The [plugin](https://gitlab.com/RoPP/flake8-use-pathlib) has not been updated for two years. We've extended the plugin rules before using the `PTH2XX` code. This

### PR DESCRIPTION
**Pull request by InSyncWithFoo**
## Summary

Resolves #14490.

## Test Plan

`cargo nextest run` and `cargo insta test`.

### CODE REVIEW

**Comment by MichaReiser** [comment]
I noticed two common patterns when reviewing the ecosystem checks:

```
if os.listdir("dir"): ....

if "file" in os.listdir("dir"):
```

The first requires using `len`, `list`, or `any` because using the `Path.iterdir` directly always returns true. 

The second mainly becomes more verbose. I'm interested in more opinions if we should exclude them. Wdyt @sbrugman 

https://github.com/apache/airflow/blob/440c224af5592f9007eef43d1dbe9025aa34e177/docs/build_docs.py#L456

https://github.com/bokeh/bokeh/blob/829b2a75c402d0d0abd7e37ff201fbdfd949d857/examples/server/app/simple_hdf5/main.py#L19
https://github.com/zulip/zulip/blob/65f05794ee59d638ad054ae6602d8ebc980fb637/scripts/lib/zulip_tools.py#L657
https://github.com/zulip/zulip/blob/65f05794ee59d638ad054ae6602d8ebc980fb637/zerver/data_import/mattermost.py#L879

**Comment by sbrugman** [comment]
The pathlib rules should flag all `os.path` cases. When these rules are active I assume users made the decision to favour pathlib over `os.path`, and partially excluding some examples will be unexpected.

It could be good to already include these cases in the tests to make sure they are covered when autofix is implemented later. The complexity here is in the fix, not in the detection of the violation.

Going over the ecosystem results I realise that `os.scandir` should also be flagged (and is closer to `Path.iterdir` produces a generator). The non-trivial fixes stem from unidiomatic use of `os.path` in the first place imo.

Using `listdir` for checking that a file does not exist is even a candidate for it's own rule as this first lists _all_ files in a directory, and then only checks one:
```python
'demo_data.hdf5' not in os.listdir(app_dir)
```

Idiomatic `os.path` solution:

```python
os.path.exists(os.path.join(app_dir), 'demo_data.hdf5')
```

Checking if a directory is empty with `os.listdir` is also wasteful:
```python
not os.listdir(api_dir):
```

Users should probably write something like:

```python
next(os.scandir(api_dir), None) is None
```

**Comment by sbrugman** [comment]
No, adding a separate rule is consistent with other rules:

https://docs.astral.sh/ruff/rules/os-remove/
https://docs.astral.sh/ruff/rules/os-unlink/

The distinction could be useful for users who would like to fix `os.scandir` (since its replacement is also a generator and requires little effort), but not `os.listdir` for instance.

**Review by MichaReiser — CHANGES_REQUESTED** [review]
Thanks @sbrugman. That makes sense to me. @InSyncWithFoo happy to merge this rule once we added tests demonstrating the patterns linked in https://github.com/astral-sh/ruff/pull/14509#issuecomment-2493991373

File: crates/ruff_linter/src/rules/flake8_use_pathlib/violations.rs [hunk 1]
```diff
@@ -1103,3 +1103,48 @@ impl Violation for PyPath {
...
+///
+/// ```python
+/// p = Path(".")
+/// for d in p.iterdir():
+///     ...
+///
+/// if any(p.iterdir()):
+///     ...
+///
+/// if (p / "file").exists():
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/flake8_use_pathlib/violations.rs:1139]: This is great. Thank you

**Review by MichaReiser — APPROVED** [review]

**Pull request merged**